### ISSUE TITLE
Needs canonical examples / reference implementation (including RegExp)

### ISSUE THREAD

**Issue by coolaj86** [semver/semver.org#59]
See https://github.com/mojombo/semver/issues/32
- RegExp in various languages that correctly parse semver
- Reference implementations

JavaScript RegExp:

``` javascript
/^(([\d+)\.(\d|)\.(\d+))(?:-([\dA-Za-z\-]+(?:\.[\dA-Za-z\-]+)*))?(?:\+([\dA-Za-z\-]+(?:\.[\dA-Za-z\-]+)*))?$/
```

JavaScript Reference Implementation:

https://github.com/coolaj86/semver-utils

I would suggest having two sections - one for tested Regular Expressions that match and another for modules / reference implementations.

There should be at least canonical reference implementation of a parser / validator with an api that would make sense to copy in any language.

**Comment by ghost** [comment]
The regular expression needs to be updated to detect the presence of leading zeros.

Unfortunately, I haven't found a way to do that, otherwise I would have provided an updated example.

**Comment by coolaj86** [comment]
Can you give an example of a valid semver string that has leading zeros? and also provide a reference to the documentation that suggests this is allowed (i.e. a phrase or example in the docs)?

If so I'll add your string to the tests in semver-utils and fix it.

**Comment by ghost** [comment]
Sorry, I meant to disallow.

The regular expression as used now does not seem to reject version numbers with leading zeros..

**Comment by tbull** [comment]
We decided only recently that leading zeroes are no longer allowed: https://github.com/mojombo/semver/issues/112
That decision invalidated all regexps developed earlier.

**Comment by gvlx** [comment]
Hi,

I have been playing with the regex for version 2.0.0 and came up with this on [regex101](http://regex101.com/r/gG8cK7/1).

Expanded:

``` regex
/^
(?'MAJOR'
    0|(?:[1-9]\d*)
)
\.
(?'MINOR'
    0|(?:[1-9]\d*)
)
\.
(?'PATCH'
    0|(?:[1-9]\d*)
)
(?:-(?'prerelease'
    (?:0|(?:[1-9A-Za-z-][0-9A-Za-z-]*))
    (?:\.
        (?:0|(?:[1-9A-Za-z-][0-9A-Za-z-]*))
    )*
))?
(?:\+(?'build'
    (?:0|(?:[1-9A-Za-z-][0-9A-Za-z-]*))
    (?:\.
        (?:0|(?:[1-9A-Za-z-][0-9A-Za-z-]*))
    )*
))?
$/
```

The pre-release and build patterns are very complex because they require the 'no leading zeros' rule.

I can't figure any benefits of that over a (very) relaxed pattern as in `/[0-9A-Za-z]+([.-][0-9A-Za-z]+)*/` which is just 'dot-or-dash' separated alphanumeric identifiers (e.g "00000-aaaaa.bbbbb") which, for me, would be more useful (I usually have to use UUIDs and other mechanical identifiers).

Notice that according to the railroad diagram and the BNF (boy, isn't that hard to read! :confused:) the identifier "0000.0000.0000.0000.------" is valid (leading zeros allowed).

If you can, please supply more edge cases :smile: on the [regex101](http://regex101.com/r/gG8cK7/1) page (in the first block is all cases are valid, in the second, invalid).

Happy hacking!

**Comment by coolaj86** [comment]
I'm in the camp to veto the use of leading zeros. In JavaScript (and some other languages) parsers will default to octal and then your sorting could get all out of wack because suddenly '011' is less than '10' both lexicographically and numerically.

And what about 007 vs 07 vs 7? Numerically they're all the same in base 10 and base 8 so how would you know which version is the "newer" one?

For the love of all that is good on this earth: no leading zeros!!!

**Comment by coolaj86** [comment]
Oh, sorry I missed the part about that being a build number. I'll have to look at the spec, but I don't think the parser I mentioned prohibits this either way.

**Comment by gvlx** [comment]
Hi,

The requirement is on 2.0.0 for pre-release but not on build version:

> 9 A pre-release version (...) Numeric identifiers MUST NOT include leading zeroes. (...)

But on the [BNF](https://github.com/mojombo/semver/blob/master/semver.md#backusnaur-form-grammar-for-valid-semver-versions):

```
<pre-release identifier> ::= <alphanumeric identifier>
                           | <numeric identifier>

<build identifier> ::= <alphanumeric identifier>
                     | <digits>

<alphanumeric identifier> ::= <non-digit>
                            | <non-digit> <identifier characters>
                            | <identifier characters> <non-digit>
                            | <identifier characters> <non-digit> <identifier characters>

<identifier characters> ::= <identifier character>
                          | <identifier character> <identifier characters>

<identifier character> ::= <digit>
                         | <non-digit>

<non-digit> ::= <letter>
              | "-"

<digit> ::= "0"
          | <positive digit>
```

The railroad diagram is less clear.

So maybe the text requires some correction.
Added pull request https://github.com/mojombo/semver.org/pull/95 

So, version 2.0.1? (patterns allowed on 2.0.0 will still work here).

**Comment by gvlx** [comment]
New [regex101 pattern](http://regex101.com/r/vW1jA8/1):

```
^
(?'MAJOR'(?:
    0|(?:[1-9]\d*)
))
\.
(?'MINOR'(?:
    0|(?:[1-9]\d*)
))
\.
(?'PATCH'(?:
    0|(?:[1-9]\d*)
))
(?:-(?'prerelease'
    [0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*
))?
(?:\+(?'build'
    [0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*
))?
$
```

**Comment by FichteFoll** [comment]
> > 9 A pre-release version (...) Numeric identifiers MUST NOT include leading zeroes. (...)

~~How I read this, it means that a pre-release identifier like `0123456789` is just not interpreted as a numeric but as an alphanumeric identifier and thus compared lexically instead of numerically.~~

---

> identifiers consisting of only digits are compared numerically and identifiers with letters or hyphens are compared lexically in ASCII sort order. Numeric identifiers always have lower precedence than non-numeric identifiers.

Never mind, it appears that numeric identifiers with leading zeros are not accepted at all or at least have no precedence defined which is pretty much the same.

**Comment by haacked** [comment]
I'm cool with adding a new page that has a regex example. We just need to figure out how we'll change the layout of the site to accommodate such links. #57 has the same design issue.

**Comment by fer-rum** [comment]
The posted regex seems to have a little trouble with the spec: In [§9](http://semver.org/#spec-item-9) it states 
> Numeric identifiers MUST NOT include leading zeroes

So I assume that a version like `1.0.0-0123` should __not__ be valid; However in the provided regex it will be accepted. I suppose the error is in the _prerelease_ capture group where 
`[0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*` should be 
`[1-9A-Za-z-]+(\.[0-9A-Za-z-]+)*`.

Also, am I correct that according to spec pre-release identifiers and build metadata behave differently with respect to the leading zero policy, since §10 misses the appropriate statement? 
In this case the __build__ capture group can be reduced to 
`[0-9A-Za-z]+`, can't it?

Before I mess with the provided regex, it would be nice if someone could confirm/ falsify my suggestion.

__Edit:__ Noticed the discrepancy between pre-release and build metadata spec.

**Comment by haacked** [comment]
Good catch. However, your change would also make `1.0.0-0abc` invalid but there's no reason that shouldn't be invalid.

Taking a step back, an identifier is either a numeric or an alphanumeric. That's a bit tricky to capture in Regex.

numeric :  `[1-9][0-9*]`
alphanumeric: `[0-9]*[A-Za-z-]+[0-9A-Za-z-]*`  At least one character must be non-numeric

Hence it'd combine to be something like `([1-9][0-9*])|([0-9]*[A-Za-z-]+[0-9A-Za-z-]*)`

So complicated. 😦 Does that look correct?

**Comment by FichteFoll** [comment]
Looks good, except that the asterisk in the numeric pattern has to go out of the set.

You could also speed it up if the second set in the alpha was matched exactly once so the engine doesn't have to backtrack.

**Comment by fer-rum** [comment]
@Haacked would you like to update the regex then (Including the suggestion by @FichteFoll )?

This is probably a regurgitation of previously discussed topics, but can a reference regex be put into the spec?

**Comment by FichteFoll** [comment]
Another problem: Just `0` is not considered.

numeric : `0|[1-9][0-9]*`
alphanumeric: `[0-9]*[A-Za-z-][0-9A-Za-z-]*` at least one character must be non-numeric
combined: `0|[1-9][0-9]*|[0-9]*[A-Za-z-][0-9A-Za-z-]*`

**Comment by fer-rum** [comment]
Why should the term "0" be valid? It is purely numeric, starts with a '0' so it should be invalid. 

> Side thought:
> (Why are leading zeros in a numeric version excluded anyway? 
> Is it an explicit goal to parse the pattern as natural numbers if applicable? 
> What about negative numbers then? 
> Or hex-representation? 
> Can I introduce a leading '0' if I want the number to be interpreted as octal? 
> Specs are clear, but the intentions aren't.)

**Comment by FichteFoll** [comment]
https://en.wikipedia.org/wiki/Leading_zero

> Therefore, the usual decimal notation of integers does not use leading zeros except for the zero itself, which would be denoted as an empty string otherwise.

Leading zeroes are excluded for simplification and prevention of ambiguity. Is `0.01.1` bigger or smaller than `0.1.01`? Are they equal? If they *are* equal, why do they not have the same string representation?

For negative numbers, I can only speculate. Considering that, in numbering, you want to "start" at a certain point and increase from that onward, it makes sense to have a generally specified starting point (i.e. lowest member) as zero instead of operating on the entire set of whole numbers, which is infinite in both directions.

It should be obvious to everyone that the spec speaks of decimal numbers in all places, which are the standard numbering system pretty much everywhere in the world, afaik.

**Comment by fer-rum** [comment]
> It should be obvious to everyone that the speak speaks of decimal numbers in all places, which are the standard numbering system pretty much everywhere in the world, afaik.

The thing is that this will often be used by programming people who tend to think in octal sometimes.  :)

> Leading zeroes are excluded for simplification and prevention of ambiguity. Is 0.01.1 bigger or smaller than 0.1.01? Are they equal? If they are equal, why do they not have the same string representation?

This is not about the version numbers which are clearly decimal natural numbers w/o leading 0s. 
In the build information, lexical sorting is the way to go, which states that the first one comes before the second one and both are not identical.

I am in favour of stating explicitly that natural decimal numbers (including 0 per se) without leading 0s are the only accepted form of purely numeric notation in the build information. (Alternatively one could go the C-way and assume that pure numeric expressions with leading 0s are interpreted as an octal number.)

Hexadecimal expressions should be no problem, they are usually prefixed by `0x`, `#` or alphanumeric anyway.

I still am not aware *why exactly* the contents of the *build information* is restricted in such a way.

**Comment by jwdonahue** [comment]
Do we have an oracle of valid and invalid version strings to test against?  We should not publish any regex's without such an oracle and test scripts to verify correctness.  We should probably also do the same for version 1.0.0, which is still in use.

**Comment by jwdonahue** [comment]
@Haacked 

> We just need to figure out how we'll change the layout of the site to accommodate such links...

The current home page has a menu of version links, why not add another one with links to a FAQ and any other pages need to be added?

**Comment by haacked** [comment]
Yeah, I think we could change the menu to have the following links:

> Latest | Versions | About | FAQ

That'd probably cover what we need now. Down the road we might add "Implementations" or "Related Specs".

**Comment by wolf99** [comment]
Did get added to the site? Is there a final recommended regex?

**Comment by gvlx** [comment]
Hi,

I don't think so, although no one did advance with other options.

The current version (with the comment from @Haacked [above](https://github.com/semver/semver.org/issues/59#issuecomment-270444155)) is now this one:

```
^
    (?'MAJOR'(?:
        0|(?:[1-9]\d*)
    ))
    \.
    (?'MINOR'(?:
        0|(?:[1-9]\d*)
    ))
    \.
    (?'PATCH'(?:
        0|(?:[1-9]\d*)
    ))
    (?:-(?'prerelease'
        [1-9A-Za-z-][0-9A-Za-z-]*(\.[0-9A-Za-z-]+)*
    ))?
    (?:\+(?'build'
        [0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*
    ))?
$
```
[on regex101](https://regex101.com/r/8hsX2w/1) [online regex tester]
[on RegExr](https://regexr.com/3pjo0) [online regex tester]
[on Debuggex](https://www.debuggex.com/r/DJXnrufle_QwO7o1) [online regex railroad diagram]
[on jex.im](https://jex.im/regulex/#!flags=mg&re=%5E((%3F%3A0%7C(%3F%3A%5B1-9%5D%5Cd*)))%5C.((%3F%3A0%7C(%3F%3A%5B1-9%5D%5Cd*)))%5C.((%3F%3A0%7C(%3F%3A%5B1-9%5D%5Cd*)))(%3F%3A-(%5B1-9A-Za-z-%5D%5B0-9A-Za-z-%5D*(%5C.%5B0-9A-Za-z-%5D%2B)*))%3F(%3F%3A%5C%2B(%5B0-9A-Za-z-%5D%2B(%5C.%5B0-9A-Za-z-%5D%2B)*))%3F%24) [online regex railroad diagram]

It would be nice if someone could contribute with more tests here.

**Comment by FichteFoll** [comment]
`1.2.3-2.02123.abc` should be invalid but is matched (https://github.com/semver/semver/issues/112#issuecomment-19491971), whereas `1.0.0-0s123` is valid but not matched.

See https://github.com/semver/semver.org/issues/59#issuecomment-278940810.

**Comment by jwdonahue** [comment]
This is also invalid and matched: 99999.1.222222222+meta.cannot.be.followed.by-prerelease

**Comment by jwdonahue** [comment]
Basically, everything after the triple is (?:JustMeta|JustPrerelease|PrereleaseThenMeta).

**Comment by jwdonahue** [comment]
1.0.0+0.build.1-rc.10000aaa-kk-0.1 is a legal version string that is among the list of invalid strings in the test data. There are no special rules for numeric fields in the build meta tag. Anything in the set [0-9A-Za-z-] is allowed in any meta tag sequence.

**Comment by gvlx** [comment]
@FichteFoll, `1.0.0-0s123` is not valid because [the spec says "Numeric identifiers MUST NOT include leading zeroes."](https://semver.org/#spec-item-9).

yes, technically `0s123` is not a _full_ numeric identifier but I do think this interpretation is reasonable.

**Comment by jwdonahue** [comment]
@gvlx, 0s123 is not a numeric identifier ([0-9]), it's an alphanumeric identifier ([0-9A-Za-z-]). There is nothing reasonable about calling a cat a duck.

**Comment by jwdonahue** [comment]
It took me many tries since yesterday afternoon to forge this regex.  I've only tested it with [Expresso](http://www.ultrapico.com/expresso.htm) which is based on .NET, but these usually work everywhere I've tried porting them, with very few tweaks. Covering all the rules correctly without excessive backtracking proved to be challenging, but I think this is linear now.

@Haacked, I'll be happy to issue a PR if you can tell me where this belongs.

`^(?'Major'[0]|[1-9][0-9]*)\.(?'Minor'[0]|[1-9][0-9]*)\.(?'Patch'[0]|[1-9][0-9]*)(?'Tags'(?(\+)\+(?'Meta'[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)|(?:\-(?'Prerelease'(?:(?([0])(?([0-9]+[A-Za-z-]+[0-9A-Za-z-]*)[0-9A-Za-z-]+|[0]{1})|(?:[1-9]+[0-9]*|[0-9A-Za-z-]+))){1}(?:\.(?([0])(?([0-9]+[A-Za-z-]+[0-9A-Za-z-]*)[0-9A-Za-z-]+|[0]{1})|(?:[1-9]+[0-9]*|[0-9A-Za-z-]+)))*){1}(?:\+(?'Meta'[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))?)?))?$`

Expanded version with some comments:


```
^
(?'Major'[0]|[1-9][0-9]*)
\.
(?'Minor'[0]|[1-9][0-9]*)
\.
(?'Patch'[0]|[1-9][0-9]*)
(?'Tags'
    (?(\+) 
        (?# Starts with meta)
        \+
        (?'Meta'[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)
        |
        (?# Prerelese and optional meta)
        (?:
            \-
            (?'Prerelease'
                (?:
                    (?([0]) (?# Leading zero is either zero or part of an alphanumeric)
                        (?([0-9]+[A-Za-z-]+[0-9A-Za-z-]*)[0-9A-Za-z-]+|[0]{1})
                        |
                        (?:[1-9]+[0-9]*|[0-9A-Za-z-]+)
                    )
                ){1}
                (?:
                    \.
                    (?([0]) (?# Leading zero is either zero or part of an alphanumeric)
                        (?([0-9]+[A-Za-z-]+[0-9A-Za-z-]*)[0-9A-Za-z-]+|[0]{1})
                        |
                        (?:[1-9]+[0-9]*|[0-9A-Za-z-]+)
                    )
                )*
            ){1}
            (?:
                \+
                (?'Meta'[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)
            )?
        )?
    )
)?
$

```

Tested against this data set:

```
1.1.2-prerelease+meta
1.1.2+meta
1.1.2+meta-invalid
1.0.0-alpha
1.0.0-beta
1.0.0-alpha.beta
1.0.0-alpha.beta.1
1.0.0-alpha.1
1.0.0-alpha0.valid
1.0.0-alpha.0valid
1.0.0-alpha-a.b-c-somethinglong+build.1-aef.1-its-okay
1.0.0-rc.1+build.1
2.0.0-rc.1+build.123
0.0.4
1.2.3
10.20.30
1.2.3-beta
10.2.3-DEV-SNAPSHOT
1.2.3-SNAPSHOT-123
1.0.0
2.0.0
1.1.7
2.0.0+build.1848
2.0.1-alpha.1227
1.0.0-alpha+beta
1.2.3----RC-SNAPSHOT.12.9.1--.12+788
1.2.3----R-S.12.9.1--.12+meta
1.2.3----RC-SNAPSHOT.12.9.1--.12
1.0.0+0.build.1-rc.10000aaa-kk-0.1
99999999999999999999999.999999999999999999.99999999999999999----RC-SNAPSHOT.12.09.1--12
5.6.7-0A.0A
Begin Invalid
1.2.3-0123
1.2.3-0123.0123
1.1.2+.123
+invalid
-invalid
-invalid+invalid
-invalid.01
alpha
alpha.beta
alpha.beta.1
alpha.1
alpha+beta
alpha_beta
alpha.
alpha..
beta\
1.0.0-alpha_beta

-alpha.
1.0.0-alpha..
1.0.0-alpha..1
1.0.0-alpha...1
1.0.0-alpha....1
1.0.0-alpha.....1
1.0.0-alpha......1
1.0.0-alpha.......1
01.1.1
1.2
1.2.3.DEV
1.2-SNAPSHOT
1.2.31.2.3----RC-SNAPSHOT.12.09.1--..12+788
1.2-RC-SNAPSHOT
-1.0.3-gamma+b7718
9.8.7+meta+meta
9.8.7-whatever+meta+meta
99999999999999999999999.999999999999999999.99999999999999999----RC-SNAPSHOT.12.09.1--------------------------------..12
```

**Comment by jwdonahue** [comment]
Well apparently, all the online Java implementations don't understand the `(?(expression)Yes|No)` regex syntax. This does not bode well for an efficient and correct regex that is relatively portable. Any ideas?

**Comment by gvlx** [comment]
Thanks @jwdonahue for pointing my mistake, I had not understood there are TWO possible identifier types (numeric, and alphanumeric with "-") playing in the pre-release part, where either of them can appear multiple times, separated by a dot ".".

So the correction expression is:
```
    (?:-
        (?'prerelease'
            (?:(?:[0-9]\d*)|(?:[0-9A-Za-z-]+))
            (?:\.(?:(?:[0-9]\d*)|(?:[0-9A-Za-z-]+)))*
    ))?
```

It will take me some time to correct the above tests.
I will also take the time to find a simpler expression for non-PCRE languages.

**Comment by jwdonahue** [comment]
@gvlx,

Your adjustment fails to match this legal prerelease tag:

`----RC-SNAPSHOT.12.09.1--12`

[Expresso](http://www.ultrapico.com/expresso.htm) tells me it matches part of the line. Also, I am fairly sure that it will lead to excessive backtracking as soon as you append the meta tag code to it and add one meta tag to the end of the string. Some engines will throw an excessive backtracking exception, but .NET will just loop forever (or a very long time) trying to match it.

The key is to insure that only one expression is applied to each field of the data, otherwise, you get excessive backtracking, and that's exactly what can happen with the (?:this|that) style of expressions if any sub-expression of 'this' is also contained in 'that'.

BTW: The websites you linked to above only actually apply your regex to the first line of data in your tesd data set.  I think that may be due to the ^$ bookends without any enclosing group and quantifier.

**Comment by jwdonahue** [comment]
@gvlx 

Apologies, I tested your regex and it failed, then wrote up some long winded comment on why it might not work and then I realized it's an expression in isolation and that's really bad, so I modified it:

```
^
    (?:-
        (?'prerelease'
            (?:(?:[0-9]\d*)|(?:[0-9A-Za-z-]+))
            (?:\.(?:(?:[0-9]\d*)|(?:[0-9A-Za-z-]+)))*
    ))?
$
```

This correctly matches and fails to match on the following two test strings:

```
----RC-SNAPSHOT.12.09.1--12
----RC-SNAPSHOT.12.09.1--..12

```

But it also matches the empty line at the end, which is terribly wrong.

**Comment by jwdonahue** [comment]
This includes a fix for the above blank line anomaly, but it is definitely not linear, just add some extra dashes to the second, nearly legal test string:

```
^
    (?:-
        (?'prerelease'
            (?:(?:[0-9]\d*)|(?:[0-9A-Za-z-]+))
            (?:\.(?:(?:[0-9]\d*)|(?:[0-9A-Za-z-]+)))*
    ))+
$
```

```
----RC-SNAPSHOT.12.09.1--12
----RC-SNAPSHOT.12.09.1--------------------------------..12

```

**Comment by jwdonahue** [comment]
So, your expression is essentially exponentially complex. True, it is unlikely to matter for 99% of the real-world cases out there, but it does not scale to lists of version strings, and we need to validate and sort lists of version strings. Modern continuous integration systems can easily generate thousands of them per month.

To be fair, I am not entirely certain that my regex is always linear either. I only know that it does not take seconds or minutes to process the test data, that includes several degenerate examples.

**Comment by jwdonahue** [comment]
@gvlx, try this:

```
^
(?'Major'[0]|[1-9][0-9]*)
\.
(?'Minor'[0]|[1-9][0-9]*)
\.
(?'Patch'[0]|[1-9][0-9]*)
    (?:-
        (?'prerelease'
            (?:(?:[0-9]\d*)|(?:[0-9A-Za-z-]+))
            (?:\.(?:(?:[0-9]\d*)|(?:[0-9A-Za-z-]+)))*
    ))+
$
```

With these two test strings:

```
99999999999999999999999.999999999999999999.99999999999999999----RC-SNAPSHOT.12.09.1--12
99999999999999999999999.999999999999999999.99999999999999999----RC-SNAPSHOT.12.09.1--------------------------------..12
```

I gave up waiting on it after about 45 seconds. Pretty sure it could go on for hour or days, if it ever halts. I think the reason is all the back-tracking that happens on the nearly correct, but totally degenerate second string.

**Comment by gvlx** [comment]
Hi, @jwdonahue 

Yes, it was just a fragment and, yes, it is not terribly efficient. :/
I already knew that problem, that's the reason I warned I have to spend some more time working on it.

I just run some more tests using the debugger in regex101.com with [this example](https://regex101.com/r/8hsX2w/7) and it takes some 12 steps for each subgroup (your examples are longer).

The simplified _fragment_ for the prerelease is now (removing the non-capturing group options):
```
(-(([1-9]\d*)|([0-9a-zA-Z-]+))(\.(([1-9]\d*)|([0-9a-zA-Z-]+)))*)?
```

This seems to work on most languages.
As I don't know of an online regex checker for .Net I used your suggestion of [Expresso](http://www.ultrapico.com/expresso.htm) and it worked fine.

**Comment by jwdonahue** [comment]
To be practical, that would have to be:

```
^
(?'Major'[0]|[1-9][0-9]*)
\.
(?'Minor'[0]|[1-9][0-9]*)
\.
(?'Patch'[0]|[1-9][0-9]*)
(?'Prerelease'-(([1-9]\d*)|([0-9a-zA-Z-]+))(\.(([1-9]\d*)|([0-9a-zA-Z-]+)))*)?
(?'Meta'\+([0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))?
$
```

That gets all the valid strings in the test data, but also passes `1.2.3-0123` and `1.2.3-0123.0123`.

**Comment by jwdonahue** [comment]
Okay, I think got one that is more portable than my previous one. This is just the tag expression:

```
^(?'Tags'(?:\-(?'Prerelease'(?:(?=[0]{1}[0-9A-Za-z-]{0})(?:[0]{1})|(?=[1-9]{1}[0-9]*[A-Za-z]{0})(?:[0-9]+)|(?=[0-9]*[A-Za-z-]+[0-9A-Za-z-]*)(?:[0-9A-Za-z-]+)){1}(?:\.(?=[0]{1}[0-9A-Za-z-]{0})(?:[0]{1})|\.(?=[1-9]{1}[0-9]*[A-Za-z]{0})(?:[0-9]+)|\.(?=[0-9]*[A-Za-z-]+[0-9A-Za-z-]*)(?:[0-9A-Za-z-]+))*){1})*(?:\+(?'Meta'(?:[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)))*)$
```

Expanded version with comments:

```
^
(?'Tags'
    (?:
        \-
        (?'Prerelease'
            (?:
                (?# Look ahead, is it just zero? Then claim a zero.)
                (?=[0]{1}[0-9A-Za-z-]{0})(?:[0]{1})
                |
                (?# Look ahead, is it pure numeric? Then claim the digits.)
                (?=[1-9]{1}[0-9]*[A-Za-z]{0})(?:[0-9]+)
                |
                (?# Look ahead, is it alphanumeric? Take them all.)
                (?=[0-9]*[A-Za-z-]+[0-9A-Za-z-]*)(?:[0-9A-Za-z-]+)
            ){1}
            (?:
                (?# Look ahead, is it just zero? Then claim a zero.)
                \.(?=[0]{1}[0-9A-Za-z-]{0})(?:[0]{1})
                |
                (?# Look ahead, is it pure numeric? Then claim the digits.)
                \.(?=[1-9]{1}[0-9]*[A-Za-z]{0})(?:[0-9]+)
                |
                (?# Look ahead, is it alphanumeric? Take them all.)
                \.(?=[0-9]*[A-Za-z-]+[0-9A-Za-z-]*)(?:[0-9A-Za-z-]+)
            )*
        ){1}
    )*
    (?:\+(?'Meta'(?:[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)))*
)
$
```

Tested against this data:

```
-alpha
-alpha.beta.1
-0+meta
-12345+meta.meta
-0.abc+000567.other-meta
-0.1200005+1
-0A.0A+0A.0A
-pr.pr+whatever
-alpha
-alpha.beta.1
Invalid
-01234
-0123.0123
-0000
-0000.0000
-1234.0000
~~~

Note that this version will actually match an empty line. I'll fix that when I merge it with rest of the regex.

**Comment by jwdonahue** [comment]
Putting it all together with the latest data set:

```
^(?'Major'0|[1-9][0-9]*)\.(?'Minor'0|[1-9][0-9]*)\.(?'Patch'0|[1-9][0-9]*)(?'Tags'(?:\-(?'Prerelease'(?:(?=[0]{1}[0-9A-Za-z-]{0})(?:[0]{1})|(?=[1-9]{1}[0-9]*[A-Za-z]{0})(?:[0-9]+)|(?=[0-9]*[A-Za-z-]+[0-9A-Za-z-]*)(?:[0-9A-Za-z-]+)){1}(?:\.(?=[0]{1}[0-9A-Za-z-]{0})(?:[0]{1})|\.(?=[1-9]{1}[0-9]*[A-Za-z]{0})(?:[0-9]+)|\.(?=[0-9]*[A-Za-z-]+[0-9A-Za-z-]*)(?:[0-9A-Za-z-]+))*){1})*(?:\+(?'Meta'(?:[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))){0,1})$
```

```
^
(?'Major'0|[1-9][0-9]*)
\.
(?'Minor'0|[1-9][0-9]*)
\.
(?'Patch'0|[1-9][0-9]*)
(?'Tags'
    (?:
        \-
        (?'Prerelease'
            (?:
                (?=[0]{1}[0-9A-Za-z-]{0})(?:[0]{1})
                |
                (?=[1-9]{1}[0-9]*[A-Za-z]{0})(?:[0-9]+)
                |
                (?=[0-9]*[A-Za-z-]+[0-9A-Za-z-]*)(?:[0-9A-Za-z-]+)
            ){1}
            (?:
                \.(?=[0]{1}[0-9A-Za-z-]{0})(?:[0]{1})
                |
                \.(?=[1-9]{1}[0-9]*[A-Za-z]{0})(?:[0-9]+)
                |
                \.(?=[0-9]*[A-Za-z-]+[0-9A-Za-z-]*)(?:[0-9A-Za-z-]+)
            )*
        ){1}
    )*
    (?:\+(?'Meta'(?:[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))){0,1}
)
$
```

```
0.0.4
1.2.3
10.20.30
1.1.2-prerelease+meta
1.1.2+meta
1.1.2+meta-valid
1.0.0-alpha
1.0.0-beta
1.0.0-alpha.beta
1.0.0-alpha.beta.1
1.0.0-alpha.1
1.0.0-alpha0.valid
1.0.0-alpha.0valid
1.0.0-alpha-a.b-c-somethinglong+build.1-aef.1-its-okay
1.0.0-rc.1+build.1
2.0.0-rc.1+build.123
1.2.3-beta
10.2.3-DEV-SNAPSHOT
1.2.3-SNAPSHOT-123
1.0.0
2.0.0
1.1.7
2.0.0+build.1848
2.0.1-alpha.1227
1.0.0-alpha+beta
1.2.3----RC-SNAPSHOT.12.9.1--.12+788
1.2.3----R-S.12.9.1--.12+meta
1.2.3----RC-SNAPSHOT.12.9.1--.12
1.0.0+0.build.1-rc.10000aaa-kk-0.1
99999999999999999999999.999999999999999999.99999999999999999
Begin Invalid

1
1.2
1.2.3-0123
1.2.3-0123.0123
1.1.2+.123
+invalid
-invalid
-invalid+invalid
-invalid.01
alpha
alpha.beta
alpha.beta.1
alpha.1
alpha+beta
alpha_beta
alpha.
alpha..
beta\
1.0.0-alpha_beta
-alpha.
1.0.0-alpha..
1.0.0-alpha..1
1.0.0-alpha...1
1.0.0-alpha....1
1.0.0-alpha.....1
1.0.0-alpha......1
1.0.0-alpha.......1
01.1.1
1.01.1
1.1.01
1.2
1.2.3.DEV
1.2-SNAPSHOT
1.2.31.2.3----RC-SNAPSHOT.12.09.1--..12+788
1.2-RC-SNAPSHOT
-1.0.3-gamma+b7718
+justmeta
9.8.7+meta+meta
9.8.7-whatever+meta+meta
99999999999999999999999.999999999999999999.99999999999999999----RC-SNAPSHOT.12.09.1--------------------------------..12
```

**Comment by jwdonahue** [comment]
The problem with my last version is the [Java engine](https://regex101.com/r/8hsX2w/8) doesn't like it. I get a number of complaints about "The preceding token is not quantifiable".  Any clues?

**Comment by jwdonahue** [comment]
It is now painfully clear that the problem space is just complicated enough that it can't be solved with a single expression that work globally. Here's a PCRE/Java/.NET version that's been tested on .NET with [Expresso](http://www.ultrapico.com/expresso.htm) and [regex101.com](https://regex101.com/r/8hsX2w/9).  Seems to be linear.  It does not work with python or golang. Perhaps someone can contribute for one or both of those?

```
^(?<Major>0|[1-9][0-9]*)\.(?<Minor>0|[1-9][0-9]*)\.(?<Patch>0|[1-9][0-9]*)(?<Tags>(?:\-(?<Prerelease>(?:(?=[0]{1}[0-9A-Za-z-]{0})(?:[0]{1})|(?=[1-9]{1}[0-9]*[A-Za-z]{0})(?:[0-9]+)|(?=[0-9]*[A-Za-z-]+[0-9A-Za-z-]*)(?:[0-9A-Za-z-]+)){1}(?:\.(?=[0]{1}[0-9A-Za-z-]{0})(?:[0]{1})|\.(?=[1-9]{1}[0-9]*[A-Za-z]{0})(?:[0-9]+)|\.(?=[0-9]*[A-Za-z-]+[0-9A-Za-z-]*)(?:[0-9A-Za-z-]+))*){1}){0,1}(?:\+(?<Meta>(?:[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))){0,1})$
```
Same test data as my last post.

**Comment by gvlx** [comment]
Hi @jwdonahue , your expression works on [regex101.com](https://regex101.com/r/8hsX2w/11) with python if you remove the group naming (only a few regex engines understand that, and comments too).

It will [_almost_ works with golang](https://regex101.com/r/8hsX2w/12) if you remove the lookaheads `(?=` (that are giving the "not quantifiable" errors).

Tell me, you added some redundant matches and quantifiers `{1}`. Do they make the pattern more efficient on some regexp engines? Look at the description on [Debuggex](https://www.debuggex.com/r/t774w6ShwAfdY1D5).

**Comment by jwdonahue** [comment]
Ok, I asked Stackoverflow for help last night and got [some very useful feedback](https://stackoverflow.com/questions/50460610/more-portable-or-additional-coverage-for-semver-regex-than-this-one?noredirect=1#comment87937866_50460610). It turns out the `(?P<name>...` named group form works on PCRE, .NET, Java and Python, and there is a [PCRE library for golang](https://github.com/d4l3k/go-pcre), so this [version works on regex101.com](https://regex101.com/r/E0iVVS/2) for everything but the golang mode on regex101.com and golang coders can use it with the library.  Getting closer.

```
^(?P<VersionTripple>(?P<Major>0|[1-9][0-9]*)\.(?P<Minor>0|[1-9][0-9]*)\.(?P<Patch>0|[1-9][0-9]*)){1}(?P<Tags>(?:\-(?P<Prerelease>(?:(?=[0]{1}[0-9A-Za-z-]{0})(?:[0]{1})|(?=[1-9]{1}[0-9]*[A-Za-z]{0})(?:[0-9]+)|(?=[0-9]*[A-Za-z-]+[0-9A-Za-z-]*)(?:[0-9A-Za-z-]+)){1}(?:\.(?=[0]{1}[0-9A-Za-z-]{0})(?:[0]{1})|\.(?=[1-9]{1}[0-9]*[A-Za-z]{0})(?:[0-9]+)|\.(?=[0-9]*[A-Za-z-]+[0-9A-Za-z-]*)(?:[0-9A-Za-z-]+))*){1}){0,1}(?:\+(?P<Meta>(?:[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))){0,1})$
```

@gvlx, named capture groups are required in my book. Relying on numbered capture groups is unstable. It's too easy for a random numbered capture group to accidentally show up in a revision of the regex that breaks all the code that uses the results of the previous version. I am not just looking for match/no-match Boolean result, I want to be able to access each of the fields by name. Though it would be neat if the results could somehow have PrereleaseTags[] and MetaTags[] arrays (minus the dots), it's simple enough to split the Prerelease and Meta tags on the dots, in a post processing step.

I think we're very close now, just need to eliminate some redundant quantifiers and investigate whether there are any useful optimizations.  Regex 101 counts a total of 591 steps to process the 72 lines (1293 bytes) of test data. Not sure exactly what they are measuring there, but if there's any way to speed this thing up without losing functionality, that would be awesome.

Also, the test data is a bit random and should be rationalized and probably requires expansion.

**Comment by jwdonahue** [comment]
I have only one major worry with regard to posting a regex as a canonical example, and that is, I know of no way to prove its correctness relative to the spec.  In the absence of such a proof, we should label these as **plausible examples that are explicitly not canonical**, if we publish them at all. The spec should always be the golden standard.

**Comment by dls314** [comment]
>if we publish them at all

I think this should be published, with some of the gory details, if for no other reason than otherwise people will re-create worse regex for the same purpose.

**Comment by jwdonahue** [comment]
Yup, I just did another search and looked a bit deeper than the the first page, and there's over hundreds of them here on GitHub and most are useless and some may even be worse than useless.

**Comment by dgn** [comment]
If anyone comes here looking for a non-pcre version working in `go`, I modified jwdonahue's expression a bit to use it in my code:

```golang
var semanticVersionExpression = "^(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)(?:\\-([A-Za-z-][0-9A-Za-z-]+))?(?:\\+([0-9A-Za-z-]+))?$"
```

No guarantees given whatsoever that this has the same functionality as the original (no named subgroups, for example - but go doesn't support them anyway), but I already tested it using the original dataset, at least the negative examples.

**Comment by gvlx** [comment]
@djug, the spec says "a series of dot separated identifiers", the identifiers themselves have no full stops.

**Comment by dgn** [comment]
@gvlx I think you meant me. I removed my comment after realising I was wrong about that.

**Comment by gvlx** [comment]
Following the discussion in [EBNF grammar](https://github.com/semver/semver/issues/192) and [this clarification](https://github.com/semver/semver/issues/181), I updated this description in [W3C simple  EBNF grammar](https://www.w3.org/TR/REC-xml/#sec-notation), which is much clearer than the [official SemVer specification's BNF grammar](https://github.com/semver/semver/blob/master/semver.md#backusnaur-form-grammar-for-valid-semver-versions) to help clear the regex ambiguities:

```
Version ::= VersionCore ('-' PreRelease)? ('+' Build)?

VersionCore ::= Major '.' Minor '.' Patch
Major ::= NumericIdentifier  
Minor ::= NumericIdentifier  
Patch ::= NumericIdentifier  
PreRelease ::= CompoundPreReleaseIdentifier 
Build      ::= CompoundBuildIdentifier 

CompoundPreReleaseIdentifier ::= PreReleaseIdentifier ('.' PreReleaseIdentifier)*
CompoundBuildIdentifier      ::= BuildIdentifier      ('.' BuildIdentifier)*

PreReleaseIdentifier     ::= AlphaNumericIdentifier | NumericIdentifier+
BuildIdentifier          ::= IdentifierCharacter+

NumericIdentifier        ::= '0' | ( PositiveNumeric NumericCharacter* )
AlphaNumericIdentifier   ::= IdentifierCharacter* NonNumericCharacter IdentifierCharacter*

IdentifierCharacter ::= NonNumericCharacter | NumericCharacter

NonNumericCharacter ::= [A-Za-z-]
NumericCharacter    ::= '0' | PositiveNumeric
PositiveNumeric     ::= [1-9]

```
**Update 2018/07/17**: finally I understood (call me slooow) the construction of the pre-release identifier! The element `AlphaNumericIdentifier` is made so that it _always_ has a non-numeric character, avoiding the leading zero on pure numeric identifiers. (Thanks for @jwdonahue for correcting me on this).

Following this, I think the actual Regex for the two main extra tags are (using @DavidFichtmueller's simplifications [here](https://github.com/semver/semver/issues/232#issue-48635632)):

- BuildIdentifier ::== `/[0-9A-Za-z-]+/`
- PreReleaseIdentifier ::== `/0|[1-9]\d*|\d*[A-Za-z-][0-9A-Za-z-]*/`

which have to be tested as a _compound_ identifier, i.e., `/<identifier>(\.<identifier>)*/`.

Maybe these patterns can be further [optimized](https://github.com/semver/semver/issues/232#issuecomment-405582995)?

I suppose that, for performance reasons, pattern processing could be done on two passes?
- Extract the main groups first `/^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(-[.0-9a-zA-Z-]+)?(\+[.0-9a-zA-Z-]+)?$/` (will produce false positives);
- Test the pre-release and build matchs separately as compound identifiers

**Comment by Relequestual** [comment]
If anyone is interested, here is a [javascript regex version (regex101)](https://regex101.com/r/VOHwch/2/) I made, which passes all the same tests as those in the [above PCRE version](https://regex101.com/r/gG8cK7/1).

`^((0|[1-9][0-9]*)\.){2}(0|[1-9][0-9]*){1}(-([0-9A-Za-z-]+\.?)+)*(\+([0-9A-Za-z-]+-?\.?)+)*$`

It's a little more condensed, as it's not concerned with group names, and just validation. (I'm using it for a JSON Schema)

**Comment by Relequestual** [comment]
Question, the spec for pre-release and metadata says the identifiers may consist of ASCII alphanumerics and hyphens only. It doesn't mention anything about dots, however the previously mentioned regex allows for it. Is this implied somewhere other than by the attached examples?

**Comment by jwdonahue** [comment]
@Relequestual 

Quoting from [the spec](https://semver.org/).

> [9 . A pre-release version MAY be denoted by appending a hyphen and a series of dot separated identifiers immediately following the patch version.](https://semver.org/#spec-item-9)

> [10. Build metadata MAY be denoted by appending a plus sign and a series of dot separated identifiers immediately following the patch or pre-release version.](https://semver.org/#spec-item-10)

I would add that your regex has an unbounded or at least extremely long run-time when faced with this tortuous, nearly legal example:

99999999999999999999999.999999999999999999.99999999999999999----RC-SNAPSHOT.12.09.1--------------------------------..12

Notice the empty prerelease field just before that final '.12'? That triggers back-tracking, which can result in infinite and pointless machinations. Now it's possible that your javascript implementation simply times out and correctly fails this on your system, but it's not something you should count on in a production environment. To the best of my knowledge, .NET implementations seem to be the most susceptible (they don't give up easily). Since it can only happen with near-matches, and you simply want a pass/fail, your regex will be fine, provided that you ensure that a reasonable timeout is in place to cut-off this degenerate behavior.

**Comment by jwdonahue** [comment]
@Relequestual, FYI: I got around the back-tracking by using look-ahead to chose which regex snippet to apply. It's those sneak peaks ahead of the current anchor point that makes my regex so long. Adding named capture groups to your regex would not overly complicate it, but they won't fix the bugs in your code.

Your regex fails to reject 1.2.3-0123 and 1.2.3-0123.0123. Both are invalid due to leading zero in the purely numeric fields of the prerelease tag.

**Comment by jwdonahue** [comment]
@Relequestual, here's two more examples it failed to reject:

9.8.7+meta+meta
9.8.7-whatever+meta+meta

**Comment by jwdonahue** [comment]
@gvlx, I think I missed something in your [last post](https://github.com/semver/semver.org/issues/59#issuecomment-393560776), regarding what you are calling AlphanumIdentifierNLZ.  That terminology is incorrect. It should be something like AllNumericNoLeadingZerosIdentifier. Alphanumeric identifiers do not have the leading zero prohibition, it only applies to all numeric identifiers, so -0zyz is a legal prerelease tag according to the spec.

**Comment by FichteFoll** [comment]
Depending on your regex engine, you could also look at atomic groups or possessive quantifiers to prevent backtracking.

**Comment by gvlx** [comment]
@jwdonahue, the name comes from the [BNF](https://github.com/semver/semver/blob/master/semver.md#backusnaur-form-grammar-for-valid-semver-versions), as pointed out in https://github.com/semver/semver/issues/181#issuecomment-34962618.

Yes, not the best choice but I kept it for consistency with the [BNF](https://github.com/semver/semver/blob/master/semver.md#backusnaur-form-grammar-for-valid-semver-versions).

As you can [read](https://semver.org/#spec-item-9), it is a modified alphanumeric identifier which _cannot have a leading zero_. This only applies to the `pre-release identifier`, and it is one of the sources of performance degradation.

This is also the reason @Relequestual's https://github.com/semver/semver.org/issues/59#issuecomment-404532645 is not correct.

I still wonder if making the two step matching approach I suggested would improve performance as each regex would much simpler.

**Comment by jwdonahue** [comment]
@gvlx, are we talking past each other? Is there something I've missed? **Nothing in the spec prohibits alphanumeric identifiers from having leading zeros**. Only the pure numeric identifiers have that prohibition on them. I've been trying to get this point across to you and others for some time now. -0123zyx is a legal prerelease tag, even with the leading zero (maybe that wasn't obvious with your display font?). -0123 is not legal, due to the leading zero, -0 is legal. So -0.0cba is also legal. It's an important nuance that adds unavoidable complexity to the regex.

So again, I argue that the NLZ postfix belongs on the numeric identifier tag, not the alphanumeric identifier. In fact, the no-leading-zero numeric-identifier is a perfect description of the Major, Minor and Patch fields as well. NumericIdentifierNLZ applies to those majore, minor, patch and any numeric identifiers in the prerelease tag.

**Comment by jwdonahue** [comment]
@gvlx, I looked at atomic groups, but they aren't widely supported and weren't required to avoid the back-tracking problem and I am pretty sure that possessive quantifiers won't work due to all the cases where a zero is allowed to follow a dot.

**Comment by jwdonahue** [comment]
I am going to say this one more time, for clarity:

There are three classes of prerelease identifiers, just a zero, all numeric with no leading zero, and alphanumeric (leading zeros allowed).

**Comment by jwdonahue** [comment]
I just ran @DavidFichtmueller's suggested regex against our current data set and it passed with flying colors. See semver/semver#232.  All it lacks are named capture groups.

**Comment by Relequestual** [comment]
Thanks @jwdonahue, I guess I'll need to revise my regex!
regex101 timed out with the example you provided.
I'm not a regex expert, so open to suggestions, specifically to fix the possible performance issue.

I'm mostly expecting this to be used in cases where there are no pre-release or meta information provided, but good to note the potential issues!

**Comment by MyNameIsCosmo** [comment]
Hello all,

Back in June I played around with the regex from both @gvlx and @jwdonahue and came up with some solution for named grouping, that I further condensed down to a regex without grouping, and then a regex with POSIX /compatibility/ (I say this loosely, cause it 'worked for me')

I put the expressions in the following Github repository along with some notes:
https://github.com/T5CC/semver-regex

Here's the named capture group expression:
`^(?P<SemVer>(?P<Major>0|[1-9][0-9]*).(?P<Minor>0|[1-9][0-9]*).(?P<Patch>0|[1-9][0-9]*))(?P<Tags>(?:-(?P<Prerelease>(?:[0-9A-Za-z-]\.?)*))?(?:\+(?P<Meta>(?:[0-9A-Za-z-]\.?)*))?)?$`
https://regex101.com/r/JpUgtQ/1

Edit:
Found out that the above regex does not invalidate leading zeros in prerelease tags.

This solution hasn't been checked for every edge case, but it does cover all the ones provided by @jwdonahue and it also processes `99999999999999999999999.999999999999999999.99999999999999999----RC-SNAPSHOT.12.09.1--------------------------------..12` without recursion (though a ton of steps).

Edit:
Catching up to recent issues, https://github.com/semver/semver/issues/232 has an efficient expression.
I've updated the semver-regex repo to @DavidFichtmueller's solution, and gave credit.

**Comment by jwdonahue** [comment]
Okay, I think we've converged a good set of regex's over on semver/semver#232.  

@coolaj86 please close this thread at your earliest possible convenience.

**Comment by joewiz** [comment]
@jwdonahue Thanks for your reply to my gist and for pointing me to this issue. I've updated my semver parsing and comparing library for XQuery, which I've rewritten to use the regex and test version strings shown here. Comments welcome - though gist won't notify me of comments, so I'd appreciate a ping here or via Twitter (same username there). https://gist.github.com/joewiz/b349e2853a17bf817e5d0013d01fa9f9

**Comment by jwdonahue** [comment]
It seems that PR #460 has been completed.  We now have a pair of [well vetted pair of regex's included in the spec](https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string).

Can we please close this discussion now?
